src/main/java/net/logstash/logback/util/ReusableJsonFormatterPool.java

```diff
@@ -0,0 +1,140 @@
+
+    public ReusableJsonFormatterPool(CompositeJsonFormatter<Event> formatterFactory, int minBufferSize) {
+        this.formatterFactory = Objects.requireNonNull(formatterFactory);
+        this.minBufferSize = minBufferSize;
+    }
+
+    public class ReusableJsonFormatter implements Closeable {
+        private final ReusableByteBuffer buffer;
+        private final CompositeJsonFormatter<Event>.JsonFormatter formatter;
+        
+        ReusableJsonFormatter(ReusableByteBuffer buffer, CompositeJsonFormatter<Event>.JsonFormatter formatter) {
+            this.buffer = buffer;
+            this.formatter = formatter;
+        }
+        
+        public ReusableByteBuffer getBuffer() {
+            return buffer;
+        }
+        
+        public void write(Event event) throws IOException {
+            this.formatter.writeEvent(event);
+        }
+        
+        @Override
+        protected void finalize() throws Throwable {
+            dispose();
+        }
+        
+        public void dispose() throws IOException {
+            this.formatter.dispose();
+            this.buffer.reset();
+        }
+        
+        public void reset() {
+            this.buffer.reset();
+        }
+        
+        @Override
+        public void close() throws IOException {
+            release(this);
+        }
+    }
+    
+    public ReusableJsonFormatter acquire() throws IOException {
+        ReusableJsonFormatter cachedFormatter = null;
+
+        Deque<ReusableJsonFormatter> cachedFormatters = formatters.get();
+        if (cachedFormatters != null) {
+            cachedFormatter = cachedFormatters.poll();
+        }
+        
+        if (cachedFormatter == null) {
+            cachedFormatter = createJsonFormatter();
+        }
+        
+        return cachedFormatter;
+    }
+    
+    protected ReusableJsonFormatter createJsonFormatter() throws IOException {
+        ReusableByteBuffer buffer = new ReusableByteBuffer(this.minBufferSize);
+        CompositeJsonFormatter<Event>.JsonFormatter jsonFormatter = formatterFactory.createJsonFormatter(buffer);
+        return new ReusableJsonFormatter(buffer, jsonFormatter);
+    }
+    
+    public void release(ReusableJsonFormatter cachedFormatter) {
+        if (cachedFormatter == null) {
+            return;
+        }
+        
+//        if (isStarted()) {
+            /*
+             * Create a new Deque if the SoftReference has been cleared.
+             * The new instance may be lost if another thread executes this code concurrently and the
+             * released JsonFormatter won't be reused. This behavior is expected and is far cheaper than
+             * implementing a more robust concurrent solution.
+             */
+            Deque<ReusableJsonFormatter> cachedFormatters = formatters.get();
+            if (cachedFormatters == null) {
+                cachedFormatters = new ConcurrentLinkedDeque<>();
+                formatters = new SoftReference<>(cachedFormatters);
```

Comment:
Is `formatters` referenced from multiple threads?  If so, need to make it `volatile`

Reply:
@philsttr Is wrapping the SoftReference in an AtomicReference better?